### PR TITLE
fix(plot): not working with "xaxis=sim_time" option

### PR DESCRIPTION
**Pull request by rokamu623**
## Description

Fixed a problem with Plot not working correctly with the sim_time option.

## Related links
 - ipynb
https://github.com/tier4/CARET_analyze/files/11505189/test.zip

## Pre-review checklist for the PR author

- [x] I've confirmed the [contribution guidelines](https://github.com/tier4/caret/blob/main/.github/CONTRIBUTING.md).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] (Optional) The PR has been properly tested with CARET_report verification.
- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

**Comment by takeshi-iwanari** [comment]
@rokamu623 
With this PR, exception fault issue is fixed, but the result of `plot.show(xaxis_type='sim_time')` is not correct.
Could you check it please?

I tested with the attached trace data and script, which are created following [the tutorial for sim_time](https://tier4.github.io/CARET_doc/main/recording/sim_time/).

- The original frequency of all topics are 10 Hz
- The rosbag is played with `-r 0.2`
- message_flow works as expected
    - When sim_time=False, the frequency of topics are 2 Hz
    - When sim_time=True, the frequency of topics are 10 Hz. and beginning time is the same as rosbag
- However, create_frequency_timeseries_plot doesn't work as expected
    - When sim_time=False, the frequency of topics are 2 Hz
    - When sim_time=True, the frequency of topics are 2 Hz. **It's should be 10 Hz !!!**
    
image
image
[test.zip](https://github.com/tier4/CARET_analyze/files/11166818/test.zip)

**Review by atsushi421 — APPROVED** [review]
@rokamu623 
LGTM.
The figure was drawn correctly in `xaxis_type='sim_time'`.
- [ipynb](https://drive.google.com/drive/folders/1p7HYZghXPku3_JhSsgh129hc2ihsltOd?usp=share_link)
- [trace data](https://drive.google.com/file/d/1HFIOV_nk_rc-sq7yCPxp-XQM9jsZvo33/view?usp=share_link)

However, the value of the z-axis labels is too large when using simtime, which eliminates the difference between the displayed times.
This issue may need to be addressed in a separate ticket.

**Comment by atsushi421** [comment]
@takeshi-iwanari 
The bug was caused by calculating the frequency before converting to simtime (same for the latency and the period).
https://github.com/tier4/CARET_analyze/blob/e68cc31932ce07f77d4bb26975b0b839be6a8a2c/src/caret_analyze/plot/timeseries/frequency_timeseries.py#L102

@rokamu623
Can I ask for an implementation fix to address this bug?

**Comment by rokamu623** [comment]
Problem solved with conversion to `sim_time` after `Frequency` calculation.
Please review once. @atsushi421 @takeshi-iwanari 

This has resulted in correct behaviour when `sim_time` is specified in `Plot.create_frequency_timeseries_plot`.

Fixes are in progress for period and latency.

**Review by takeshi-iwanari — COMMENTED** [review]
@rokamu623 
Thank you for merging main (https://github.com/tier4/CARET_analyze/pull/285). I confirmed `Plot.create_frequency_timeseries_plot()` works as expected.
(I don't approve this PR at the moment because `Plot.create_period_timeseries_plot(()` and `Plot.create_latency_timeseries_plot()` need to be fixed as well)

**Comment by tamegaictc2** [comment]
@rokamu623
I fixed the source code and updated the ipynb file.
Could you please review and confirm?
Screenshot from 2023-05-17 15-20-37
[test.zip](https://github.com/tier4/CARET_analyze/files/11494959/test.zip)

**Comment by rokamu623** [comment]
@tamegaictc2 
1. Commit conflict has occurred.
   - Please merge `main` branch into `rokamu623:fix/bug_sim_time` branch to resolve the conflict.

2. When you commit, please remember to add the --signoff option.

3. Please include the results without the `xaxis="sim_time"` option for better understanding.

**Comment by rokamu623** [comment]
@tamegaictc2

1. Commit conflict has occurred.

2. Please merge `main` branch into `rokamu623:fix/bug_sim_time` branch to resolve the conflict.
   - When you commit, please remember to add the `--signoff` option.

3. Please include the results without the `xaxis="sim_time"` option for better understanding.

**Comment by tamegaictc2** [comment]
@rokamu623 
After making the necessary fixes in the commit.(f7058603810036b1080ed17c2e98753a4d614e64)
I retryed test the ipynb file.
please confirm?
Screenshot from 2023-05-18 15-50-02
[test.zip](https://github.com/tier4/CARET_analyze/files/11505189/test.zip)

**Comment by rokamu623** [comment]
@atsushi421 
This PR appears to be causing conflict with #291 .
Is it safe to merge it as it is?

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py [hunk 1]
```diff
@@ -66,20 +67,51 @@ def create_figure(self) -> Figure:
         # Apply xaxis offset
         records_range = Range([to.to_records() for to in target_objects])
         frame_min, frame_max = records_range.get_range()
```
Comment (rokamu623) [inline src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:69]:
```suggestion
        x_range_name = None
        frame_min, frame_max = records_range.get_range()
```

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py [hunk 2]
```diff
@@ -66,20 +67,51 @@ def create_figure(self) -> Figure:
...
+                        converter_cb = comm.callback_subscription
+                        converter = converter_cb._provider.get_sim_time_converter()
+                        break
+            else:
+                converter = target_objects[0]._provider.get_sim_time_converter()
+        if converter:
+            frame_min_convert = converter.convert(frame_min)
+            frame_max_convert = converter.convert(frame_max)
+            x_range_name = 'x_plot_axis'
+            apply_x_axis_offset(fig, frame_min_convert, frame_max_convert, x_range_name)
```
Comment (rokamu623) [inline src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:87]:
```suggestion
            frame_min = converter.convert(frame_min)
            frame_max = converter.convert(frame_max)
            x_range_name = 'x_plot_axis'
            apply_x_axis_offset(fig, frame_min, frame_max, x_range_name)
```

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py [hunk 3]
```diff
@@ -66,20 +67,51 @@ def create_figure(self) -> Figure:
...
         # Draw lines
         color_selector = ColorSelectorFactory.create_instance(coloring_rule='unique')
         legend_manager = LegendManager()
-        line_source = LineSource(legend_manager, target_objects[0], frame_min, self._xaxis_type)
+        if converter:
+            line_source = \
+                LineSource(legend_manager, target_objects[0], frame_min_convert, self._xaxis_type)
+        else:
+            line_source = \
+                LineSource(legend_manager, target_objects[0], frame_min, self._xaxis_type)
```
Comment (rokamu623) [inline src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:99]:
```suggestion
        line_source = LineSource(legend_manager, target_objects[0], frame_min, self._xaxis_type)
```

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py [hunk 4]
```diff
@@ -66,20 +67,51 @@ def create_figure(self) -> Figure:
...
+                    source=line_source.generate(to, timeseries),
+                    color=color_selector.get_color(),
+                    x_range_name=x_range_name
+                )
+            else:
+                renderer = fig.line(
+                    'x', 'y',
+                    source=line_source.generate(to, timeseries),
+                    color=color_selector.get_color()
+                )
```
Comment (rokamu623) [inline src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:114]:
```suggestion
            renderer = fig.line(
                'x', 'y',
                source=line_source.generate(to, timeseries),
                color=color_selector.get_color(),
                x_range_name=x_range_name
            )
```

**Review by rokamu623 — COMMENTED** [review]
@atsushi421 @tamegaictc2 
The coding appears redundant.
How about the following?

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 4 above)
Comment (tamegaictc2) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:114]: Reflecting the changes mentioned above, result of the test.ipynb file is below.
`ValueError: failed to validate GlyphRenderer(id='1047', ...).x_range_name: expected a value of type str, got None of type NoneType`

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 1 above)
Comment (tamegaictc2) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:69]: Please refer to the following for details.
https://github.com/tier4/CARET_analyze/pull/276#discussion_r1204935711

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 2 above)
Comment (tamegaictc2) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:87]: The variables "frame_min" and "frame_max" are of type "int," and the return value of "converter.convert()" is of type "float." An error occurred in "test.mypy." How about the following modifications?
```
            frame_min = int(converter.convert(frame_min))
            frame_max = int(converter.convert(frame_max))
            x_range_name = 'x_plot_axis'
            apply_x_axis_offset(fig, frame_min, frame_max, x_range_name)
```

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 3 above)
Comment (tamegaictc2) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:99]: I accept this proposal.

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 4 above)
Comment (rokamu623) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:114]: @tamegaictc2 
It appears that x_range_name is a specification that does not accept None.

Only the part of `renderer = ...` should be implemented with if else.

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 1 above)
Comment (rokamu623) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:69]: From the difference between the return values of `records_range.get_range()` and `converter.convert()`, it is better to have another variable for `flame_min/max` and `flame_min/max_contert`.

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 2 above)
Comment (rokamu623) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:87]: From the difference between the return values of `records_range.get_range()` and `converter.convert()`, it is better to have another variable for `flame_min/max` and `flame_min/max_contert`.

Hence, there is no need to apply the above suggestion.

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 3 above)
Comment (rokamu623) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:99]: From the difference between the return values of `records_range.get_range()` and `converter.convert()`, it is better to have another variable for `flame_min/max` and `flame_min/max_contert`.

Hence, there is no need to apply the above suggestion.

File: src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py (hunk 4 above)
Comment (rokamu623) [reply src/caret_analyze/plot/visualize_lib/bokeh/timeseries.py:114]: From the difference between the return values of `records_range.get_range()` and `converter.convert()`, it is better to have another variable for `flame_min/max` and `flame_min/max_contert`.

Hence, there is no need to apply the above suggestion.

**Comment by rokamu623** [comment]
@atsushi421 @nabetetsu 
As review from me have been completed, please review it.

**Review by atsushi421 — DISMISSED** [review]
LGTM!

**Comment by nabetetsu** [comment]
I was changing the rate of `ros2 bag play` command to see if it works with same results, but the median period results displayed in the graph created by `Plot.create_period_timeseries_plot()` appear to be different.
Please give me a few moments to check it out deeply.

**Comment by rokamu623** [comment]
@nabetetsu 
Problems with changing the rate of `ros2 bag play` will be treated in another PR.
We would like to close this PR on the condition that it does not cause syntax errors.

**Review by nabetetsu — APPROVED** [review]
image
I have confirmed that I can create graph with simtime option.

note: spell-check-differential error will be fixed on https://github.com/tier4/CARET_analyze/pull/304.

**Pull request merged**